---
title: "CRISPRi_analysis"
author: "Puwen Tan"
date: "2021/6/10"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. pre-processing
```{r}
library(dplyr)
raw_count <- read.table("../ext_data/bencnmark_data/CRISPRi/raw/CRISPRi_count_mat.tsv",sep="\t",header=T,check.names=F)
raw_colData <- read.table("../ext_data/bencnmark_data/CRISPRi/CRISPRi_colData.tsv",sep="\t",header=T,check.names=F)
row.names(raw_colData) <- raw_colData$cell
# 1. filter samples with at least 10 cells 
filter_sample <- raw_colData %>% group_by(guide_RNA) %>% summarise(num_cells=length(cell)) %>% filter(num_cells>=10)
col_data <- raw_colData %>% filter(guide_RNA %in% filter_sample$guide_RNA)
raw_count <- raw_count[,col_data$cell]

# 2. get differential expressed samples

get_diff_TF <- function(pos_samp, 
                        neg_samp,
                        count_mat, 
                        col_data){
  pos_cells <- col_data[col_data$guide_RNA %in% pos_samp,"cell"]
  neg_cells <- col_data[col_data$guide_RNA %in% neg_samp,"cell"]
  TF <- unique(col_data[pos_cells,"TF"])
  count_mat <- as.matrix(count_mat)
  if(!(TF %in% row.names(count_mat))) stop("No TF found in count matrix!")
  counts <- count_mat[TF,c(pos_cells, neg_cells),drop=TRUE]
  mean_pos <- mean(counts[pos_cells])
  mean_ctrl <- mean(counts[neg_cells])
  logFC <- log2(mean_pos)-log2(mean_ctrl)
  pval <- t.test(counts[pos_cells], counts[neg_cells])$p.value
  data.frame(sample=pos_samp,mean_pos,mean_ctrl,logFC,pval)
}

ctrl_samps <- col_data %>% filter(TF=="Scramble") %>% pull(guide_RNA) %>% unique()
TFs <- col_data %>% filter(TF!="Scramble") %>% pull(TF) %>% unique()
TFs <- intersect(TFs,row.names(count_mat))
ko_samp_list <- col_data %>% filter(TF %in% TFs) %>% pull(guide_RNA) %>% unique()
diff_TF_tab <- lapply(ko_samp_list,get_diff_TF,neg_samp=ctrl_samps,count_mat=raw_count,col_data=col_data)
diff_TF_tab <- data.table::rbindlist(diff_TF_tab)
diff_TF_sig <- diff_TF_tab %>% filter(pval < 0.05 & logFC < -1)
col_data_filter <- col_data %>% filter(guide_RNA %in% c(diff_TF_sig$sample, ctrl_samps))
fwrite(as.data.table(col_data_filter),"../ext_data/bencnmark_data/CRISPRi/CRISPRi_colData.tsv",sep="\t")

count_mat_filter <- count_mat[,col_data_filter$cell]
count_mat_filter <- as.data.table(count_mat_filter,keep.rownames = TRUE)
setnames(count_mat_filter,"rn","gene_name")
fwrite(count_mat_filter,"../ext_data/bencnmark_data/CRISPRi/CRISPRi_count_mat.tsv",sep="\t")
```

## 2. load data and create object
```{r}
library(data.table)
count_mat_in <- fread("../ext_data/bencnmark_data/CRISPRi/CRISPRi_count_mat.tsv",sep="\t",header=T,data.table = F)
count_mat <- count_mat_in[,-1]
row.names(count_mat) <- count_mat_in[,1]
col_data <- read.table("../ext_data/bencnmark_data/CRISPRi/CRISPRi_colData.tsv",sep="\t",header=T)
row.names(col_data) <- col_data$cell

library(dplyr)
regulons <- dorothea_regulons$hs_regulons
regulons <- dfToList(df = regulons, tf_col = "tf",target_col = "target")
# filter regulons
regulons <- regulons[Reduce(intersect,list(row.names(count_mat),names(regulons),col_data$TF))]
atfr <- scATFR(exp_data = count_mat,col_data = col_data, regulons = regulons)
```

## 3. get Activate TF regulons

```{r}
atfr <- readRDS("../ext_data/CRISPRi/diff_cells/grns.rds")

#---run GRNs
run_time <- list()
for (i in c("sclink", "pcor", "spearman","genie3")){
  st <- Sys.time()
  atfr <- filterRegulons(x=atfr, use_grn=i, ncores=16)
  diff_time <- Sys.time()-st
  run_time[[i]] <- diff_time
}

atfr <- filterRegulons(x=atfr, ncores=4)

run_time <- list()
for (j in c("viper", "aucell", "ssgsea")){
  st <- Sys.time()
  atfr <- regulonActivity(atfr, method=j,ncores=4)
  diff_time <- Sys.time()-st
  run_time[[j]] <- diff_time
}

```
## 4. ROC

```{r}
atfr <- readRDS("../ext_data/CRISPRi/GRNs/atfr.rds")
get_tf_score <- function(tf,mat,col_data,control="Scramble",type=c('KO','OE')){
  type <- match.arg(type)
  mat <- as.matrix(mat)
  tf <- Reduce(intersect,list(tf,row.names(mat),col_data[['TF']]))
  if(length(tf)!=1) stop("Only input one TF each time!")
  tf_cells <- col_data[col_data$TF %in% tf,"cell"]
  ctrl_cells <- col_data[col_data$TF %in% control,"cell"]
  all_cells <- c(tf_cells,ctrl_cells)
  active_score <- mat[tf,all_cells,drop=TRUE]
  #---
  pos_label <- rep(1,length(tf_cells))
  names(pos_label) <- tf_cells
  neg_label <- rep(-1,length(ctrl_cells))
  names(neg_label) <- ctrl_cells
  all_label <- c(pos_label, neg_label)[all_cells]
  if(type=="KO"){
    active_score <- -1*active_score
  }
  list(prediction=active_score[all_cells],label=all_label)
}
#---get ROC
get_rocr <- function(active_list,usetf=NULL, measure="auc",...){
  if(is.null(names(active_list))) stop("No names in active_list")
  if(!is.null(usetf)){
    active_list <- active_list[intersect(usetf,names(active_list))]
  }
  tfs <- names(active_list)
  prediction <- lapply(active_list,function(x){x[["prediction"]]})
  label <- lapply(active_list,function(x){x[["label"]]})
  pred <- ROCR::prediction(predictions = prediction, labels = label,...)
  perf <- ROCR::performance(pred, measure,...)
  if(length(perf@x.values)==length(tfs)){
    names(perf@x.values) <- tfs
  }
  if(length(perf@y.values)==length(tfs)){
    names(perf@y.values) <- tfs
  }
  perf
}
viper_mat <- assay(altExp(atfr,"aucell"))
viper_mat <- assay(altExp(atfr,"viper"))
viper_mat <- assay(altExp(atfr,"ssgsea"))
col_data <- colData(atfr)

TFs <- intersect(unique(col_data$TF),row.names(viper_mat))

TF_score_list <- lapply(TFs,get_tf_score,mat=viper_mat, col_data=col_data)
names(TF_score_list) <- TFs
perf <- get_rocr(TF_score_list, measure = c("tpr", "fpr"))
perf_auc <- get_rocr(TF_score_list, measure = c("auc"))

viper_auc <- unlist(perf_auc@y.values)
aucell_auc <- unlist(perf_auc@y.values)
auc_frm <- data.frame(viper=viper_auc, aucell=aucell_auc)



plot(perf,
     avg='vertical',
     spread.estimate='stderror',
     lwd=3,main='Vertical averaging + 1 standard error',
     col='blue')


plot(perf,
     colorize=TRUE,
     lwd=2,
     main='ROC curves from 10-fold cross-validation')

p <- plot(perf, avg = "threshold", colorize=TRUE,
     lwd= 3,
     main= "With ROCR you can produce standard plots\nlike ROC curves ...")
plot(perf,
     lty=3,
     col="grey78",
     add=TRUE)

```
### 5.SCENIC flow

```{r}
library(GENIE3)
library(RcisTarget)
library(AUCell)
library(data.table)

hs_motifranking <- importRankings("../ext_data/cisTarget_databases/hg38__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather")
hs_annotation <- fread("../ext_data/cisTarget_databases/motifs-v9-nr.hgnc-m0.001-o0.0.tbl",sep="\t")
mm_motifranking <- importRankings("../ext_data/cisTarget_databases/mm10__refseq-r80__500bp_up_and_100bp_down_tss.mc9nr.feather")
mm_annotation <- fread("../ext_data/cisTarget_databases/motifs-v8-nr.mgi-m0.001-o0.0.tbl",sep="\t")

#atfr <- readRDS("../ext_data/Nkx2-1_KO/GRNs/GRNs_all_atfr.rds")
#---get regulons
exprMatr <- assay(atfr)
grn_tab <- GRN(atfr,"genie3")
grn_frm <- GENIE3::getLinkList(as.matrix(grn_tab),threshold = 0)
top_num <- round(0.05*ncol(grn_tab))
grn_frm <- grn_frm %>% group_by(regulatoryGene) %>% top_n(top_num,weight)
grn_list <- split(grn_frm$targetGene,as.character(grn_frm$regulatoryGene))

#---run cistarget
grn_cistarget_res <- cisTarget(geneSets = grn_list,motifRankings = hs_motifranking, motifAnnot = hs_annotation, highlightTFs=row.names(grn_tab),nesThreshold = 0)
dir.create("../ext_data/CRISPRi/diff_cells/SCENIC/")
saveRDS(grn_cistarget_res,"../ext_data/CRISPRi/diff_cells/SCENIC/grn_cistarget_all.rds")
grn_cistarget <- grn_cistarget_res[TFinDB=="**",][enrichedGenes!="",]

library(stringr)
filterTFlists <- function(str){
  str <- str_replace_all(str, "\\s\\(directAnnotation\\)\\.",";")
  str <- str_replace_all(str, "\\s\\(inferredBy_Orthology\\)\\.",";")
  str <- str_replace_all(str, "\\s\\(inferredBy_MotifSimilarity\\)\\.",";")
  str <- str_replace_all(str, "\\s\\(inferredBy_MotifSimilarity_n_Orthology\\)\\.",";")
  str <- str_replace_all(str, "\\;\\s?+$","")
  str <- str_split(str,pattern = "\\;\\s")
  return(str)
}
grn_cistarget$TF_highConf <- filterTFlists(grn_cistarget$TF_highConf)
grn_cistarget$TF_lowConf <- filterTFlists(grn_cistarget$TF_lowConf)

grn_cistarget_filter <- grn_cistarget[apply(grn_cistarget,1,function(x){
  x['geneSet'] %in% x['TF_highConf']
}),]
grn_cistarget_filter$enrichedGenes <- str_split(grn_cistarget_filter$enrichedGenes,pattern =";")
grn_regulon <- split(grn_cistarget_filter$enrichedGenes,grn_cistarget_filter$geneSet)
grn_regulon <- lapply(grn_regulon,function(x){unique(unlist(x))})
grn_regulon_num <- sapply(grn_regulon,length)
grn_regulon <- grn_regulon[names(grn_regulon_num[which(grn_regulon_num>4)])]
#---get expression ranks
cells_rankings <- AUCell_buildRankings(exprMatr, nCores=1, plotStats=FALSE)
cells_AUC <- AUCell_calcAUC(geneSets=grn_regulon, rankings = cells_rankings, normAUC = TRUE)
auc_value <- getAUC(cells_AUC)
saveRDS(auc_value,"../ext_data/CRISPRi/diff_cells/SCENIC/AUC_value.rds")

#------performance
mat <- auc_value

i="ARNT"

TFs <- row.names(mat)
auroc <- list()
for(i in TFs){
  col_data_i <- col_data[col_data$TF %in% c(i,'Scramble'),]
  prediction <- mat[i,row.names(col_data_i ),drop=TRUE]*(-1)
  labels <- col_data_i[names(prediction),"TF"]
  label <- ifelse(labels=="Scramble",1,-1)
  pred <- ROCR::prediction(predictions = prediction, labels = label)
  perf <- ROCR::performance(pred,"auc")
  auroc[[i]] <- unlist(perf@y.values)
}
SCENIC_AUC <- unlist(auroc)
auc_frm$scenic <- SCENIC_AUC[row.names(auc_frm)]

labels <- colData(atfr)[['orig.ident']]
names(labels) <- row.names(colData(atfr))
label <- ifelse(labels=="WT",1,-1)
prediction <- mat[1,names(labels)]*(-1)
pred <- ROCR::prediction(predictions = prediction, labels = label)
perf <- ROCR::performance(pred,"auc")
perf@y.values
perf <- ROCR::performance(pred,"tpr", "fpr")

plot(perf,
     avg="threshold",
     spread.estimate="boxplot")

```




