---
title: "R Notebook"
output: html_notebook
---

### 1. integration of databases TF-targets

#### 1.1 TRANSFACT

```{r}
library(tidyverse)
TRANSFACT_TF_Tar <- read_delim("D:/HSC_Autophagy/important_genes/TFs/TRANSFAC Curated Transcription Factor Targets/gene_attribute_edges.txt",delim = "\t",col_names = T) %>% select(target,source,weight)
TRANSFACT_TF_Tar <- unique(TRANSFACT_TF_Tar[-1,])
colnames(TRANSFACT_TF_Tar) <- c("tf","target","weight")
dir.create("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database",recursive = TRUE)
write_delim(TRANSFACT_TF_Tar,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database/TRANSFAC_curated_network.txt",delim = "\t")
```

#### 1.2 ENCODE
```{r}
ENCODE_TF_Tar <- read_delim("D:/HSC_Autophagy/important_genes/TFs/ENCODE Transcription Factor Targets/gene_attribute_edges.txt",delim = "\t",col_names = T) %>% select(target,source,weight)
ENCODE_TF_Tar <- unique(ENCODE_TF_Tar[-1,])
colnames(ENCODE_TF_Tar) <- c("tf","target","weight")
dir.create("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based",recursive = TRUE)
write_delim(ENCODE_TF_Tar,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based/ENCODE_curated_network.txt",delim = "\t")

```

#### 1.3 CHEA
```{r}
CHEA_TF_Tar <- read_delim("D:/HSC_Autophagy/important_genes/TFs/CHEA Transcription Factor Targets/gene_attribute_edges.txt",delim = "\t",col_names = T) %>% select(target,source,weight)
CHEA_TF_Tar <- unique(CHEA_TF_Tar[-1,])
colnames(CHEA_TF_Tar) <- c("tf","target","weight")
dir.create("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based",recursive = TRUE)
write_delim(CHEA_TF_Tar,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based/CHEA_curated_network.txt",delim = "\t")

```
#### 1.4 merge network from msigdb

```{r}
msigdb_tf <- fgsea::gmtPathways("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/c3.tft.v7.4.symbols.gmt")
tfs <- names(msigdb_tf)
msigdb_tbl <- map2_df(.x = tfs,.y = msigdb_tf,.f = function(x,y){
  data.frame(tf=rep(x,length(y)),target=y,weight=1)
})
msigdb_tf <- str_split(msigdb_tbl$tf,pattern = "_")
msigdb_tbl$tf <- sapply(msigdb_tf,function(x){x[1]})
tf_genes <- read.table("D:/HSC_Autophagy/3.TFKO_data/Human_TFs/TF_names_v_1.01.txt",sep="\t",header=F)
tf_genes <- tf_genes$V1
msigdb_tbl <- msigdb_tbl %>% filter(tf %in% tf_genes)
write_delim(msigdb_tbl,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database/MSIGDB_curated_network.txt",delim = "\t")

```

#### 1.5 data from 2019-gr

```{r}
#---database
file_path <- 'D:/HSC_Autophagy/3.TFKO_data/2019_geno_res/rebuilt.data1/data/TF_target_sources/curated_databases'
out_dir <- 'D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database/'
dbs <- list.files(file_path)
dbs_filter <- dbs[-c(5,9,13,15,16,19:21)]
lapply(dbs_filter,function(x,in_dir,out_dir){
  tab <- read_delim(file.path(in_dir,x,"network.sif"),delim = "\t",col_names = TRUE)
  if(ncol(tab)==2){
    tab$weight <- rep(1,nrow(tab))
  }
  colnames(tab) <- c("tf","target","weight")
  write_delim(tab,file.path(out_dir,paste(str_to_upper(x),"_curated_network.txt",sep="")),delim = "\t")
},in_dir=file_path,out_dir=out_dir)

```


#### 1.6 2012 Neph TF-TF interaction
```{r}
Neph2012_dir <- "D:/HSC_Autophagy/4.TF-Target_data/2012Neph/cell6312mmc4"
cells <- list.dirs(Neph2012_dir,full.names = F)
cells <- cells[cells!=""]
merge_tab <- lapply(cells,function(x,in_dir){
  tab <- read.table(file.path(in_dir,x,"genes-regulate-genes.txt"),sep="\t",header=F)
},in_dir=Neph2012_dir)
merge_df <- do.call(rbind,merge_tab) %>% unique(.)
colnames(merge_df) <- c("tf","target")
merge_df$weight <- rep(1,nrow(merge_df))
write_delim(merge_df,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database/Neph2012_curated_network.txt",delim = "\t")

```

#### 1.7 2016 Marbach

```{r}
Marbach2016_dir <- "D:/HSC_Autophagy/4.TF-Target_data/Marbach2016/32_high-level_networks"
Marbach2016_files <- list.files(Marbach2016_dir)

Marbach2016_networks <- lapply(Marbach2016_files,function(x, in_dir){
  tab <- read_delim(file.path(in_dir,x),delim = "\t",col_names = FALSE)
  colnames(tab) <- c("tf","target","weight")
  tab$weight <- (tab$weight-min(tab$weight))/(max(tab$weight)-min(tab$weight))
  write_delim(tab,file.path("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/Marbach2016",x),delim="\t",col_names = TRUE)
  tab
},in_dir=Marbach2016_dir)

```

#### 1.8 TF KO data

```{r}
TF_KO_tbl <- read_delim("D:/HSC_Autophagy/3.TFKO_data/KnockTF/differential expression of genes in all datasets.txt",delim = "\t",col_names = TRUE)
tf_genes <- read.table("D:/HSC_Autophagy/3.TFKO_data/Human_TFs/TF_names_v_1.01.txt",sep="\t",header=F)
TF_KO_filter_sample <- TF_KO_tbl %>% filter(TF == Gene & P_value<0.1 & Log2FC<0 & up_down==2 & TF %in% tf_genes$V1) %>% pull(Sample_ID) %>% unique(.)

TF_KO_tbl_filter <- TF_KO_tbl %>% filter(Sample_ID %in% TF_KO_filter_sample & up_down!=0 & P_value<0.05 & `Mean Expr. of Control` >0 & abs(Log2FC)>1)
TF_KO_tbl_filter$`-log10P` <- -log10(TF_KO_tbl_filter$P_value)
TF_KO_tbl_list <- split(TF_KO_tbl_filter,TF_KO_tbl_filter$Sample_ID)
saveRDS(TF_KO_tbl_list,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/TF_KO_data/KnockTF_filter.rds")
TF_KO_tbl_list_out <- lapply(TF_KO_tbl_list,function(x){
  tab <- x[,c("TF","Gene","-log10P")]
  colnames(tab) <- c("tf","target","weight")
  tab$weight <- (tab$weight-min(tab$weight))/(max(tab$weight)-min(tab$weight))
  tab
})

TF_KO_tbl_list_merge <- integraNets(grnNets = TF_KO_tbl_list_out, keep_neg_regu = FALSE)
write_delim(TF_KO_tbl_list_merge,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/TF_KO_data/KnockTF_network.txt",delim="\t",col_names = TRUE)

```


#### 1.9 merge database source network

```{r}
database_dir <- "D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database"
files <- list.files(database_dir)

curated_network <- lapply(files,function(x,in_dir){
  tab <- read.delim(file.path(in_dir,x),sep="\t",header=T)
},in_dir=database_dir)

curated_merged_network <- integraNets(grnNets = curated_network, keep_neg_regu = FALSE)

write.table(curated_merged_network,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database_network.txt", sep="\t",row.names=F,quote=F)

```

#### 1.10 TFBS predicting

```{r}
#----TFBS
jaspar_2018_predict <- read_delim("D:/HSC_Autophagy/3.TFKO_data/2019_geno_res/rebuilt.data1/data/TF_target_sources/TFBS_scanning/jaspar_2018/network_n1000.sif",delim = "\t",col_names = T)
colnames(jaspar_2018_predict) <- c("tf","target")
jaspar_2018_predict$weight <- rep(1,nrow(jaspar_2018_predict))
write_delim(jaspar_2018_predict,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based/jaspar_2018_TFBS_predicting_network.txt",delim = "\t")
hocomoco_v11_predict <- read_delim("D:/HSC_Autophagy/3.TFKO_data/2019_geno_res/rebuilt.data1/data/TF_target_sources/TFBS_scanning/hocomoco_v11/network_n1000.sif",delim = "\t",col_names = T)
colnames(hocomoco_v11_predict) <- c("tf","target")
hocomoco_v11_predict$weight <- rep(1,nrow(hocomoco_v11_predict))
write_delim(hocomoco_v11_predict,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based/hocomoco_v11_TFBS_predicting_network.txt",delim = "\t")
```

#### 1.11 chip-seq based

```{r}
chipseq_dir <- "D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based"
files <- list.files(chipseq_dir)
chipseq_network <- lapply(files,function(x,in_dir){
  tab <- read.delim(file.path(in_dir,x),sep="\t",header=T)
  tab
},in_dir=chipseq_dir)
chipseq_merged_network <- integraNets(grnNets = chipseq_network, keep_neg_regu = FALSE)
write.table(chipseq_merged_network,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based_network.txt", sep="\t",row.names=F,quote=F)
```

### 2. merge all types of networks
#### 2.1 merge by Marbach2016

```{r}
net_dir <- "D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data"
chip_net <- read_delim(file.path(net_dir,"chipseq_based_network.txt"),delim = "\t",col_names = T)
database_net <- read_delim(file.path(net_dir,"curated_database_network.txt"),delim = "\t",col_names = T)
TFKO_net <- read_delim(file.path(net_dir,"KnockTF_network.txt"),delim = "\t",col_names = T)
merge_net <- integraNets(grnNets = list(chip_net, database_net, TFKO_net), keep_neg_regu = FALSE)

#------
Marbach2016_dir <- "D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/Marbach2016"
Marbach2016_files <- list.files(Marbach2016_dir)

Marbach2016_merged_network_by_tissue <- lapply(Marbach2016_files,function(x, net, in_dir){
  tab <- read_delim(file.path(in_dir,x),delim = "\t",col_names = TRUE)
  merge_tab <- integraNets(grnNets = list(tab,net),norm_weight = FALSE, keep_neg_regu = FALSE)
  write_delim(merge_tab,file.path("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/Merged_network",x),delim="\t",col_names = TRUE)
merge_tab
},net=merge_net,in_dir=Marbach2016_dir)

RUNX1_top <- merge_tab %>% filter(tf=="RUNX1") %>% arrange(desc(weight))



```

#### 2.2 merge by tissue specific

```{r}
# get database net
database_net <- read.table("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/curated_database_network.txt",sep="\t",header=T)
chipseq_net <- read.table("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/chipseq_based_network.txt",sep="\t",header=T)
tfko_net <- read.table("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/KnockTF_network.txt",sep="\t",header=T)
merge_net <- integraNets(list(database_net, chipseq_net, tfko_net),norm_weight = FALSE,keep_neg_regu = FALSE)

GTEx_dir <- "D:/HSC_Autophagy/3.TFKO_data/2019_geno_res/rebuilt.data1/data/TF_target_sources/inferred/GTEx/tissue_specific"
GTEx_tissue <- list.dirs(GTEx_dir,full.names = FALSE)[-1]
tf_genes <- read.table("D:/HSC_Autophagy/3.TFKO_data/Human_TFs/TF_names_v_1.01.txt",sep="\t",header=F)
tf_genes <- tf_genes$V1
GTEx_merged_network <- lapply(GTEx_tissue,function(x, net, tf_genes, in_dir){
  tab <- read_delim(file.path(in_dir,x,"network.txt"),delim = "\t",col_names = TRUE)
  merge_tab <- integraNets(grnNets = list(tab, net), norm_weight = TRUE, keep_neg_regu = FALSE)
  merge_tab <- merge_tab %>% filter(tf %in% tf_genes)
  write_delim(merge_tab,file.path("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/Merged_GTEx_network",paste(x,".txt.gz",sep="")),delim="\t", col_names = TRUE)
merge_tab
},net=merge_net, in_dir=GTEx_dir,tf_genes=tf_genes)
names(GTEx_merged_network) <- GTEx_tissue

tissue_specific_net <- GTEx_merged_network
saveRDS(tissue_specific_net,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/GTEx_network.rds")

#---pantissue network
pantissue_network <- read_delim("D:/HSC_Autophagy/3.TFKO_data/2019_geno_res/rebuilt.data1/data/TF_target_sources/inferred/GTEx/pantissue/network_i2.sif",delim ="\t")
pantissue_network$effect <- rep(1,nrow(pantissue_network))
pantissue_merge_network <- integraNets(grnNets = list(pantissue_network, merge_net), norm_weight = TRUE, keep_neg_regu = FALSE)
pantissue_merge_network <- pantissue_merge_network %>% filter(tf %in% tf_genes)
write_delim(pantissue_merge_network,"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/pantissue_merge_network.txt.gz",delim="\t",col_names = TRUE)


```

#### 2.3. get homolog genes


```{r}
pantissue_network <- read_delim("D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/pantissue_merge_network.txt.gz",delim = "\t",col_names = T)
mouse_homolog <- getHomologGene(input_gene = unique(c(pantissue_network$tf,pantissue_network$target)), output_type = "gene_symbol")
pantissue_network_mm <- pantissue_network %>% left_join(x = .,y = mouse_homolog,by=c("tf"="Gene_Symbol_human")) %>%
  left_join(x = .,y = mouse_homolog,by=c("target"="Gene_Symbol_human"))
pantissue_network_mm <- pantissue_network_mm[,c(4,5,3)]
colnames(pantissue_network_mm) <- c("tf","target","weight")
pantissue_network_mm  <- pantissue_network_mm %>% distinct(.) %>% na.omit(.)

write_delim(unique(pantissue_network_mm),"D:/HSC_Autophagy/Markdown/3.scATFR/ext_data/TF_Target_data/mouse/mm_pantissue_network.txt.gz",delim="\t",col_names = TRUE)

```



